### PR TITLE
Fix the problem of Workflow terminates after parallel tasks execution, merge node not triggered

### LINKED ISSUE #12492: Workflow terminates after parallel tasks execution, merge node not triggered
### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

0.14.2

### Cloud or Self Hosted

Self Hosted (Source)

### Steps to reproduce

I made a simple workflow test case,
image
When two IF-ELSE nodes execute in parallel and their results are merged into a subsequent node, if the input satisfies the IF condition of both nodes simultaneously, the subsequent node will not be executed.
image

I tried to fix this by myself, but the graph engine is too complicated.

### ✔️ Expected Behavior

The workflow executes properly.

### ❌ Actual Behavior

The workflow stops early.

### PR DESCRIPTION
**Pull request by lazyFrogLOL**
# Summary

Fix https://github.com/langgenius/dify/issues/12492

# Description
When both if conditions are satisfied, the subsequent node is executed incompletely due to the following reasons:
1. When init a Graph object, the _recursively_add_parallels method is used to create the mapping for parallel branches. When a node is found to have multiple outgoing edges, it is identified as the starting point of a parallel branch.
2. In the GraphEngine's _run method, when multiple outgoing edges are encountered:
```
else:
    parallel_generator = self._run_parallel_branches(
        edge_mappings=edge_mappings,
        in_parallel_id=in_parallel_id,
        parallel_start_node_id=parallel_start_node_id,
        handle_exceptions=handle_exceptions,
    )
```
_run_parallel_branches creates a new thread for each branch to execute. 

However:
* When one branch completes execution, it places the result in the queue.
* Once all branches have completed execution, the next node is executed.
* If multiple branches satisfy the conditions and point to the same target node, only the first branch to complete will actually execute the target node.

## Major Modifications:

1. Collect all target nodes from the branches.
2. For multiple branches pointing to the same target node, check which branches satisfy the conditions.
4. Ensure that all branches meeting the conditions can trigger the execution of the target node.

This ensures that when multiple if condition branches are satisfied, their shared subsequent nodes are executed correctly.

# Screenshots

| Before | After |
|--------|-------|
| image| image   |

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

### CODE REVIEW

File: api/core/workflow/graph_engine/graph_engine.py [hunk 1]
```diff
@@ -439,83 +439,94 @@ def _run_parallel_branches(
...
-        if not parallel_id:
-            node_id = edge_mappings[0].target_node_id
-            node_config = self.graph.node_id_config_mapping.get(node_id)
-            if not node_config:
+        
+        target_nodes: dict[str, list[GraphEdge]] = {}
+        for edge in edge_mappings:
+            if edge.target_node_id not in target_nodes:
+                target_nodes[edge.target_node_id] = []
+            target_nodes[edge.target_node_id].append(edge)
```
Comment (yihong0618) [inline api/core/workflow/graph_engine/graph_engine.py]: this can use defaultdict?

Comment (lazyFrogLOL) [reply api/core/workflow/graph_engine/graph_engine.py]: Sure, using defaultdict would be better.

```
from collections import defaultdict

target_nodes = defaultdict(list)
target_nodes[edge.target_node_id].append(edge)
```
I will commit this updating asap.

**Comment by lazyFrogLOL** [comment]
I found that this issue still exists in some scenarios and has not been fully resolved. Please hold off on merging this pull request.

**Comment by lazyFrogLOL** [comment]
The new code passes all graph engine unit tests and has been verified in several corner cases tested on the web app. Please review the code thoroughly and engage in a detailed discussion to ensure that the issue is truly resolved without breaking any existing functionality.

**Comment by yihong0618** [comment]
> The new code passes all graph engine unit tests and has been verified in several corner cases tested on the web app. Please review the code thoroughly and engage in a detailed discussion to ensure that the issue is truly resolved without breaking any existing functionality.

amazing work!

**Comment by yihong0618** [comment]
you need to fix the mypy typehint

**Comment by lazyFrogLOL** [comment]
> you need to fix the mypy typehint

thx for the reminder.

**Comment by lazyFrogLOL** [comment]
> > you need to fix the mypy typehint
> 
> thx for the reminder.

Done the file reformating and lint fixing.

**Comment by Nov1c444** [comment]
Thank you very much for your contribution. Here, we try not to change the logic of the graph_engine. The main reason for the problem is that this line of code does not meet the condition. (graph_engine.py: 432)
```python
if in_parallel_id and self.graph.node_parallel_mapping.get(next_node_id, "") != in_parallel_id:
                break
``` 
Therefore, it would be more elegant to directly modify the logic in the graph.(graph.py: 594) :)
```python
 if (
                        branch_node_id != branch_node_id2
                        and node_id in inner_route2
                        and len(reverse_edge_mapping.get(node_id, [])) > 1
                        and cls._is_node_in_routes(
                            reverse_edge_mapping=reverse_edge_mapping,
                            start_node_id=node_id,
                            routes_node_ids=routes_node_ids,
                        )
                        # Exclude conditional branch nodes
                        and all(edge.run_condition is None for edge in reverse_edge_mapping.get(node_id, []))
                    ):
```

**Comment by lazyFrogLOL** [comment]
> Thank you very much for your contribution. Here, we try not to change the logic of the graph_engine. The main reason for the problem is that this line of code does not meet the condition. (graph_engine.py: 432)
> 
> ```python
> if in_parallel_id and self.graph.node_parallel_mapping.get(next_node_id, "") != in_parallel_id:
>                 break
> ```
> 
> Therefore, it would be more elegant to directly modify the logic in the graph.(graph.py: 594) :)
> 
> ```python
>  if (
>                         branch_node_id != branch_node_id2
>                         and node_id in inner_route2
>                         and len(reverse_edge_mapping.get(node_id, [])) > 1
>                         and cls._is_node_in_routes(
>                             reverse_edge_mapping=reverse_edge_mapping,
>                             start_node_id=node_id,
>                             routes_node_ids=routes_node_ids,
>                         )
>                         # Exclude conditional branch nodes
>                         and all(edge.run_condition is None for edge in reverse_edge_mapping.get(node_id, []))
>                     ):
> ```

Really impressive! I’m looking forward to seeing this solution deployed in the new version.

**Review by Nov1c444 — APPROVED** [review]

**Comment by Kevin9703** [comment]
This change could cause the order of the output content to become jumbled in certain scenarios. When I commented out this code block, the output returned to normal.
image
image

**Comment by Kevin9703** [comment]
> > This change could cause the order of the output content to become jumbled in certain scenarios. When I commented out this code block, the output returned to normal. image image
> 
> Could you provide me with a DSL to reproduce this issue?

[test.txt](https://github.com/user-attachments/files/18762546/test.txt)
